File: src/router.py
"""Request routing and forwarding logic."""

import json
import time
from typing import Any, AsyncIterator

import httpx

from .config import ConfigManager, get_config_manager
from .models import (
    MessagesRequest,
    MessagesResponse,
    ModelConfig,
    ErrorResponse,
    ErrorDetail,
)


class RouterError(Exception):
    """Router related errors."""

    def __init__(self, message: str, status_code: int = 500, error_type: str = "router_error"):
        super().__init__(message)
        self.message = message
        self.status_code = status_code
        self.error_type = error_type

    def to_error_response(self) -> ErrorResponse:
        """Convert to an ErrorResponse model."""
        return ErrorResponse(
            error=ErrorDetail(type=self.error_type, message=self.message)
        )


class ModelRouter:
    """Routes requests to appropriate model backends."""

    def __init__(self, config_manager: ConfigManager | None = None):
        """Initialize the router.

        Args:
            config_manager: ConfigManager instance. If None, uses global instance.
        """
        self.config_manager = config_manager or get_config_manager()
        self._client: httpx.AsyncClient | None = None

    async def _get_client(self) -> httpx.AsyncClient:
        """Get or create the HTTP client."""
        if self._client is None or self._client.is_closed:
            timeout = self.config_manager.config.gateway.timeout
            self._client = httpx.AsyncClient(
                timeout=httpx.Timeout(timeout, connect=30.0),
                follow_redirects=True,
            )
        return self._client

    async def close(self) -> None:
        """Close the HTTP client."""
        if self._client and not self._client.is_closed:
            await self._client.aclose()
            self._client = None

    def resolve_route(self, model_name: str) -> tuple[str, ModelConfig]:
        """Resolve model name to route configuration.

        Args:
            model_name: The model name from the request

        Returns:
            Tuple of (resolved_model_name, ModelConfig)

        Raises:
            RouterError: If model is not found
        """
        # Resolve alias if needed
        resolved_name = self.config_manager.resolve_model_name(model_name)
        model_config = self.config_manager.get_model(resolved_name)

        if not model_config:
            raise RouterError(
                f"Model '{model_name}' not found. Available models: "
                f"{', '.join(self.config_manager.config.models.keys())}",
                status_code=400,
                error_type="invalid_model",
            )

        return resolved_name, model_config

    def build_headers(
        self,
        model_config: ModelConfig,
        original_headers: dict[str, str],
    ) -> dict[str, str]:
        """Build headers for the upstream request.

        Args:
            model_config: The model configuration
            original_headers: Headers from the original request

        Returns:
            Headers dict for the upstream request

        Raises:
            RouterError: If API key is not available
        """
        # Get API key
        api_key = self.config_manager.get_api_key(
            self.config_manager.resolve_model_name(model_config.model_id)
        )

        # Try to find the key by iterating models
        if not api_key:
            for name, config in self.config_manager.config.models.items():
                if config.model_id == model_config.model_id or config == model_config:
                    api_key = self.config_manager.get_api_key(name)
                    if api_key:
                        break

        if not api_key:
            raise RouterError(
                f"API key not configured for model '{model_config.display_name}'. "
                f"Please set the {model_config.api_key_env} environment variable.",
                status_code=401,
                error_type="authentication_error",
            )

        headers = {
            "Content-Type": "application/json",
            "Accept": "application/json",
            model_config.auth_header: api_key,
        }

        # Add Anthropic-specific headers for Anthropic provider
        if model_config.provider == "anthropic":
            headers["anthropic-version"] = model_config.api_version
            # Forward anthropic-beta header if present
            if "anthropic-beta" in original_headers:
                headers["anthropic-beta"] = original_headers["anthropic-beta"]

        # Add any extra headers from config
        headers.update(model_config.extra_headers)

        return headers

    def build_request_body(
        self,
        request: MessagesRequest,
        model_config: ModelConfig,
    ) -> dict[str, Any]:
        """Build the request body for the upstream API.

        Args:
            request: The original request
            model_config: The target model configuration

        Returns:
            Request body dict
        """
        # Start with the request data
        body = request.model_dump(exclude_none=True)

        # Replace model name with the actual model ID
        body["model"] = model_config.model_id

        # Ensure max_tokens doesn't exceed model limit
        if body.get("max_tokens", 0) > model_config.max_tokens:
            body["max_tokens"] = model_config.max_tokens

        return body

    def build_url(self, model_config: ModelConfig, endpoint: str = "/v1/messages") -> str:
        """Build the full URL for the upstream API.

        Args:
            model_config: The model configuration
            endpoint: The API endpoint path

        Returns:
            Full URL string
        """
        base_url = model_config.base_url.rstrip("/")

        # Handle different API structures
        # Anthropic: https://api.anthropic.com/v1/messages
        # Others might have different paths

        if model_config.provider == "anthropic":
            return f"{base_url}{endpoint}"

        # For other providers, the base_url usually includes the necessary path
        # e.g., https://api.deepseek.com/anthropic already points to the right place
        if endpoint == "/v1/messages":
            # Most Anthropic-compatible APIs use /v1/messages from their base
            if "/anthropic" in base_url or "/coding" in base_url or "/apps/anthropic" in base_url:
                return f"{base_url}/v1/messages"
            return f"{base_url}{endpoint}"

        return f"{base_url}{endpoint}"

    async def forward_request(
        self,
        request: MessagesRequest,
        original_headers: dict[str, str],
    ) -> MessagesResponse:
        """Forward a non-streaming request to the appropriate backend.

        Args:
            request: The messages request
            original_headers: Original request headers

        Returns:
            The response from the upstream API

        Raises:
            RouterError: On routing or API errors
        """
        resolved_name, model_config = self.resolve_route(request.model)
        headers = self.build_headers(model_config, original_headers)
        body = self.build_request_body(request, model_config)
        url = self.build_url(model_config)

        # Ensure stream is False for non-streaming
        body["stream"] = False

        client = await self._get_client()
        start_time = time.time()

        try:
            response = await client.post(url, json=body, headers=headers)
            latency_ms = (time.time() - start_time) * 1000

            if response.status_code != 200:
                error_body = response.text
                try:
                    error_json = response.json()
                    error_message = error_json.get("error", {}).get("message", error_body)
                except Exception:
                    error_message = error_body

                raise RouterError(
                    f"Upstream API error ({model_config.provider}): {error_message}",
                    status_code=response.status_code,
                    error_type="api_error",
                )

            response_data = response.json()

            # Log request info
            if self.config_manager.config.gateway.enable_logging:
                usage = response_data.get("usage", {})
                print(
                    f"[{resolved_name}] Request completed in {latency_ms:.0f}ms | "
                    f"Input: {usage.get('input_tokens', 'N/A')} | "
                    f"Output: {usage.get('output_tokens', 'N/A')}"
                )

            return MessagesResponse(**response_data)

        except httpx.TimeoutException:
            raise RouterError(
                f"Request to {model_config.provider} timed out after "
                f"{self.config_manager.config.gateway.timeout}s",
                status_code=504,
                error_type="timeout_error",
            )
        except httpx.RequestError as e:
            raise RouterError(
                f"Failed to connect to {model_config.provider}: {str(e)}",
                status_code=502,
                error_type="connection_error",
            )

    async def forward_stream(
        self,
        request: MessagesRequest,
        original_headers: dict[str, str],
    ) -> AsyncIterator[str]:
        """Forward a streaming request to the appropriate backend.

        Args:
            request: The messages request
            original_headers: Original request headers

        Yields:
            SSE formatted strings

        Raises:
            RouterError: On routing or API errors
        """
        resolved_name, model_config = self.resolve_route(request.model)
        headers = self.build_headers(model_config, original_headers)
        body = self.build_request_body(request, model_config)
        url = self.build_url(model_config)

        # Ensure stream is True
        body["stream"] = True

        # Update Accept header for SSE
        headers["Accept"] = "text/event-stream"

        client = await self._get_client()
        start_time = time.time()

        try:
            async with client.stream("POST", url, json=body, headers=headers) as response:
                if response.status_code != 200:
                    error_body = await response.aread()
                    try:
                        error_json = json.loads(error_body)
                        error_message = error_json.get("error", {}).get("message", error_body.decode())
                    except Exception:
                        error_message = error_body.decode()

                    # Yield error as SSE event
                    error_event = {
                        "type": "error",
                        "error": {"type": "api_error", "message": error_message},
                    }
                    yield f"event: error\ndata: {json.dumps(error_event)}\n\n"
                    return

                # Stream the response
                buffer = ""
                async for chunk in response.aiter_text():
                    buffer += chunk

                    # Process complete SSE events
                    while "\n\n" in buffer:
                        event_str, buffer = buffer.split("\n\n", 1)
                        if event_str.strip():
                            yield event_str + "\n\n"

                # Don't forget any remaining data
                if buffer.strip():
                    yield buffer + "\n\n"

                # Log completion
                if self.config_manager.config.gateway.enable_logging:
                    latency_ms = (time.time() - start_time) * 1000
                    print(f"[{resolved_name}] Stream completed in {latency_ms:.0f}ms")

        except httpx.TimeoutException:
            error_event = {
                "type": "error",
                "error": {
                    "type": "timeout_error",
                    "message": f"Request timed out after {self.config_manager.config.gateway.timeout}s",
                },
            }
            yield f"event: error\ndata: {json.dumps(error_event)}\n\n"

        except httpx.RequestError as e:
            error_event = {
                "type": "error",
                "error": {
                    "type": "connection_error",
                    "message": f"Connection error: {str(e)}",
                },
            }
            yield f"event: error\ndata: {json.dumps(error_event)}\n\n"


# Global router instance
_router: ModelRouter | None = None


def get_router() -> ModelRouter:
    """Get the global router instance."""
    global _router
    if _router is None:
        _router = ModelRouter()
    return _router


async def shutdown_router() -> None:
    """Shutdown the global router."""
    global _router
    if _router:
        await _router.close()
        _router = None

File: src/router.ts
/**
 * Request routing and forwarding logic
 */

// Uses the Node 18+ built-in fetch: node-fetch v3 is ESM-only and breaks
// the CJS dist on Node <22.12 (ERR_REQUIRE_ESM).
import type { ConfigManager } from './config.js';
import type { UsageTracker } from './usage.js';
import type {
  MessagesRequest,
  MessagesResponse,
  ErrorResponse,
  ModelConfig,
  RouteInfo,
} from './types.js';

export class RouterError extends Error {
  statusCode: number;
  errorType: string;

  constructor(message: string, statusCode = 500, errorType = 'router_error') {
    super(message);
    this.name = 'RouterError';
    this.statusCode = statusCode;
    this.errorType = errorType;
  }

  toErrorResponse(): ErrorResponse {
    return {
      type: 'error',
      error: {
        type: this.errorType,
        message: this.message,
      },
    };
  }
}

/** Upstream failures worth retrying on a fallback model. */
function isRetryable(error: RouterError): boolean {
  return (
    error.errorType === 'connection_error' ||
    error.errorType === 'timeout_error' ||
    error.statusCode === 429 ||
    error.statusCode >= 500
  );
}

function parseUpstreamErrorMessage(errorText: string): string {
  try {
    const errorJson = JSON.parse(errorText);
    return errorJson.error?.message || errorText;
  } catch {
    return errorText;
  }
}

interface StreamUsageAccumulator {
  input: number;
  output: number;
}

export class ModelRouter {
  private configManager: ConfigManager;
  private usageTracker?: UsageTracker;

  constructor(configManager: ConfigManager, usageTracker?: UsageTracker) {
    this.configManager = configManager;
    this.usageTracker = usageTracker;
  }

  private isLoggingEnabled(): boolean {
    return this.configManager.getConfig().gateway.enable_logging;
  }

  resolveRoute(modelName: string): RouteInfo {
    const resolvedName = this.configManager.resolveModelName(modelName);
    const modelConfig = this.configManager.getModel(resolvedName);

    if (!modelConfig) {
      const availableModels = Object.keys(this.configManager.getConfig().models).join(', ');
      throw new RouterError(
        `Model '${modelName}' not found. Available models: ${availableModels}`,
        400,
        'invalid_model'
      );
    }

    const apiKey = this.configManager.getApiKey(resolvedName);
    if (!apiKey) {
      throw new RouterError(
        `API key not configured for model '${modelConfig.display_name}'. ` +
          `Please set the ${modelConfig.api_key_env} environment variable.`,
        401,
        'authentication_error'
      );
    }

    return {
      name: resolvedName,
      config: modelConfig,
      apiKey,
    };
  }

  /**
   * The primary model followed by its configured fallbacks (aliases
   * resolved, duplicates removed). Fallbacks are one level deep by design:
   * a fallback's own fallback list is ignored, so chains cannot cycle.
   */
  buildChain(modelName: string): string[] {
    const primary = this.configManager.resolveModelName(modelName);
    const chain = [primary];
    const fallbacks = this.configManager.getModel(primary)?.fallback ?? [];
    for (const fallback of fallbacks) {
      const resolved = this.configManager.resolveModelName(fallback);
      if (!chain.includes(resolved)) {
        chain.push(resolved);
      }
    }
    return chain;
  }

  buildHeaders(route: RouteInfo, originalHeaders: Record<string, string>): Record<string, string> {
    const authHeader = route.config.auth_header || 'x-api-key';
    const authType =
      route.config.auth_type || (authHeader.toLowerCase() === 'authorization' ? 'bearer' : 'api_key');
    const authValue = authType === 'bearer' ? `Bearer ${route.apiKey}` : route.apiKey;
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
      Accept: 'application/json',
      [authHeader]: authValue,
    };

    // Add Anthropic-specific headers for Anthropic provider
    if (route.config.provider === 'anthropic') {
      headers['anthropic-version'] = '2023-06-01';
      if (originalHeaders['anthropic-beta']) {
        headers['anthropic-beta'] = originalHeaders['anthropic-beta'];
      }
    }

    return headers;
  }

  buildRequestBody(request: MessagesRequest, modelConfig: ModelConfig): Record<string, unknown> {
    const body = { ...request };

    // Replace model name with actual model ID
    body.model = modelConfig.model_id;

    // Cap max_tokens if exceeds model limit
    const maxTokens = modelConfig.max_tokens || 8192;
    if (body.max_tokens && body.max_tokens > maxTokens) {
      body.max_tokens = maxTokens;
    }

    this.normalizeProviderRequestBody(body, modelConfig);

    return body;
  }

  private normalizeProviderRequestBody(body: Record<string, unknown>, modelConfig: ModelConfig): void {
    if (modelConfig.provider === 'deepseek') {
      this.normalizeDeepSeekRequestBody(body);
    }
  }

  private normalizeDeepSeekRequestBody(body: Record<string, unknown>): void {
    // DeepSeek documents Anthropic `metadata` as ignored, but the endpoint can still
    // reject Claude Code session metadata when `user_id` contains unsupported chars.
    delete body.metadata;
    delete body.user_id;
  }

  buildUrl(modelConfig: ModelConfig, endpoint = '/v1/messages'): string {
    let baseUrl = modelConfig.base_url.replace(/\/$/, '');

    // Handle different API structures
    if (modelConfig.provider === 'anthropic') {
      return `${baseUrl}${endpoint}`;
    }

    // For other providers with /anthropic or /coding in base URL
    if (
      baseUrl.includes('/anthropic') ||
      baseUrl.includes('/coding') ||
      baseUrl.includes('/apps/anthropic')
    ) {
      return `${baseUrl}/v1/messages`;
    }

    return `${baseUrl}${endpoint}`;
  }

  async forwardRequest(
    request: MessagesRequest,
    originalHeaders: Record<string, string>
  ): Promise<MessagesResponse> {
    const chain = this.buildChain(request.model);
    let lastError: RouterError | null = null;

    for (let i = 0; i < chain.length; i++) {
      let route: RouteInfo;
      try {
        route = this.resolveRoute(chain[i]);
      } catch (error) {
        if (i === 0) {
          // Primary resolution failures are client misconfiguration - no failover
          throw error;
        }
        continue; // Skip fallbacks that are unknown or missing a key
      }

      try {
        const response = await this.attemptRequest(route, request, originalHeaders);
        this.usageTracker?.record(route.name, response.usage ?? {});

        if (this.isLoggingEnabled()) {
          console.log(
            `[${route.config.display_name}] Request completed | ` +
              `Input: ${response.usage?.input_tokens ?? 'N/A'} | ` +
              `Output: ${response.usage?.output_tokens ?? 'N/A'}`
          );
        }
        return response;
      } catch (error) {
        const routerError =
          error instanceof RouterError
            ? error
            : new RouterError('Unknown error occurred', 500, 'internal_error');
        this.usageTracker?.recordError(route.name);
        lastError = routerError;

        const hasMoreCandidates = i < chain.length - 1;
        if (isRetryable(routerError) && hasMoreCandidates) {
          if (this.isLoggingEnabled()) {
            console.log(
              `[failover] ${route.name} failed (${routerError.message}), trying ${chain[i + 1]}`
            );
          }
          continue;
        }
        throw routerError;
      }
    }

    throw lastError ?? new RouterError('No usable model in fallback chain', 500, 'internal_error');
  }

  private async attemptRequest(
    route: RouteInfo,
    request: MessagesRequest,
    originalHeaders: Record<string, string>
  ): Promise<MessagesResponse> {
    const headers = this.buildHeaders(route, originalHeaders);
    const body = this.buildRequestBody(request, route.config);
    const url = this.buildUrl(route.config);

    // Ensure stream is false for non-streaming
    body.stream = false;

    const timeout = this.configManager.getConfig().gateway.timeout * 1000;
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), timeout);

    try {
      const response = await fetch(url, {
        method: 'POST',
        headers,
        body: JSON.stringify(body),
        signal: controller.signal,
      });

      if (!response.ok) {
        const errorText = await response.text();
        throw new RouterError(
          `Upstream API error (${route.config.provider}): ${parseUpstreamErrorMessage(errorText)}`,
          response.status,
          'api_error'
        );
      }

      return (await response.json()) as MessagesResponse;
    } catch (error) {
      if (error instanceof RouterError) {
        throw error;
      }
      if (error instanceof Error) {
        if (error.name === 'AbortError') {
          throw new RouterError(`Request timed out after ${timeout / 1000}s`, 504, 'timeout_error');
        }
        throw new RouterError(`Connection error: ${error.message}`, 502, 'connection_error');
      }
      throw new RouterError('Unknown error occurred', 500, 'internal_error');
    } finally {
      clearTimeout(timeoutId);
    }
  }

  async *forwardStream(
    request: MessagesRequest,
    originalHeaders: Record<string, string>
  ): AsyncGenerator<string> {
    const chain = this.buildChain(request.model);
    const timeout = this.configManager.getConfig().gateway.timeout * 1000;
    let lastError: RouterError | null = null;

    for (let i = 0; i < chain.length; i++) {
      let route: RouteInfo;
      try {
        route = this.resolveRoute(chain[i]);
      } catch (error) {
        if (i === 0) {
          const routerError =
            error instanceof RouterError
              ? error
              : new RouterError('Unknown error occurred', 500, 'internal_error');
          yield this.formatErrorEvent(routerError.errorType, routerError.message);
          return;
        }
        continue;
      }

      const headers = this.buildHeaders(route, originalHeaders);
      const body = this.buildRequestBody(request, route.config);
      const url = this.buildUrl(route.config);
      body.stream = true;
      headers.Accept = 'text/event-stream';

      const controller = new AbortController();
      const connectTimer = setTimeout(() => controller.abort(), timeout);
      let response: Response;

      try {
        response = await fetch(url, {
          method: 'POST',
          headers,
          body: JSON.stringify(body),
          signal: controller.signal,
        });
      } catch (error) {
        clearTimeout(connectTimer);
        const routerError =
          error instanceof Error && error.name === 'AbortError'
            ? new RouterError(`Request timed out after ${timeout / 1000}s`, 504, 'timeout_error')
            : new RouterError(
                `Connection error: ${error instanceof Error ? error.message : 'Unknown error'}`,
                502,
                'connection_error'
              );
        this.usageTracker?.recordError(route.name);
        lastError = routerError;
        if (i < chain.length - 1) {
          this.logFailover(route.name, routerError.message, chain[i + 1]);
          continue;
        }
        yield this.formatErrorEvent(routerError.errorType, routerError.message);
        return;
      }
      clearTimeout(connectTimer);

      if (!response.ok) {
        const errorText = await response.text();
        const routerError = new RouterError(
          parseUpstreamErrorMessage(errorText),
          response.status,
          'api_error'
        );
        this.usageTracker?.recordError(route.name);
        lastError = routerError;
        if (isRetryable(routerError) && i < chain.length - 1) {
          this.logFailover(route.name, routerError.message, chain[i + 1]);
          continue;
        }
        yield this.formatErrorEvent('api_error', routerError.message);
        return;
      }

      // Connected: stream the body. No failover past this point - the
      // client may already have received partial output.
      yield* this.streamBody(route, response, controller, timeout);
      return;
    }

    const message = lastError?.message ?? 'No usable model in fallback chain';
    yield this.formatErrorEvent(lastError?.errorType ?? 'internal_error', message);
  }

  private async *streamBody(
    route: RouteInfo,
    response: Response,
    controller: AbortController,
    timeoutMs: number
  ): AsyncGenerator<string> {
    if (!response.body) {
      this.usageTracker?.recordError(route.name);
      yield this.formatErrorEvent('internal_error', 'No response body');
      return;
    }

    const usage: StreamUsageAccumulator = { input: 0, output: 0 };
    let buffer = '';
    let eventCount = 0;

    // Idle watchdog: a stream that stops producing chunks for the full
    // gateway timeout is aborted instead of hanging the connection forever.
    let lastChunkAt = Date.now();
    const idleTimer = setInterval(() => {
      if (Date.now() - lastChunkAt > timeoutMs) {
        controller.abort();
      }
    }, 1000);

    try {
      for await (const value of response.body) {
        lastChunkAt = Date.now();
        const chunk = typeof value === 'string' ? value : Buffer.from(value).toString('utf-8');
        buffer += chunk;

        // Process complete SSE events
        while (buffer.includes('\n\n')) {
          const idx = buffer.indexOf('\n\n');
          const event = buffer.slice(0, idx);
          buffer = buffer.slice(idx + 2);
          if (event.trim()) {
            eventCount++;
            this.accumulateStreamUsage(event, usage);
            yield event + '\n\n';
          }
        }
      }

      // Don't forget remaining data
      if (buffer.trim()) {
        eventCount++;
        this.accumulateStreamUsage(buffer, usage);
        yield buffer + '\n\n';
      }

      this.usageTracker?.record(route.name, {
        input_tokens: usage.input,
        output_tokens: usage.output,
      });

      if (this.isLoggingEnabled()) {
        console.log(`[${route.config.display_name}] Stream completed (${eventCount} events)`);
      }
    } catch (error) {
      this.usageTracker?.recordError(route.name);

      let errorMessage = 'Unknown error';
      let errorType = 'internal_error';
      if (error instanceof Error) {
        if (error.name === 'AbortError') {
          errorMessage = `Stream idle for more than ${timeoutMs / 1000}s, aborted`;
          errorType = 'timeout_error';
        } else {
          errorMessage = error.message;
        }
      }
      yield this.formatErrorEvent(errorType, errorMessage);
    } finally {
      clearInterval(idleTimer);
    }
  }

  /** Pull token counts out of message_start / message_delta SSE events. */
  private accumulateStreamUsage(event: string, usage: StreamUsageAccumulator): void {
    if (!event.includes('"usage"')) {
      return;
    }
    const dataLine = event.split('\n').find((line) => line.startsWith('data:'));
    if (!dataLine) {
      return;
    }
    try {
      const parsed = JSON.parse(dataLine.slice(5)) as {
        type?: string;
        message?: { usage?: { input_tokens?: number; output_tokens?: number } };
        usage?: { output_tokens?: number };
      };
      if (parsed.type === 'message_start' && parsed.message?.usage) {
        usage.input += parsed.message.usage.input_tokens ?? 0;
        usage.output = parsed.message.usage.output_tokens ?? usage.output;
      } else if (parsed.type === 'message_delta' && parsed.usage) {
        // message_delta reports the cumulative output count
        usage.output = parsed.usage.output_tokens ?? usage.output;
      }
    } catch {
      // Non-JSON data lines are passed through untouched
    }
  }

  private logFailover(from: string, reason: string, to: string): void {
    if (this.isLoggingEnabled()) {
      console.log(`[failover] ${from} failed (${reason}), trying ${to}`);
    }
  }

  private formatErrorEvent(errorType: string, message: string): string {
    const errorEvent = {
      type: 'error',
      error: { type: errorType, message },
    };
    return `event: error\ndata: ${JSON.stringify(errorEvent)}\n\n`;
  }
}
